### PR TITLE
feat(infra.ci.jio): Added AZ VM templates for Ubuntu-22 AMD and ARM maven-17 and maven-21 agents

### PR DESCRIPTION
**Pull request by jayfranco999**
As per https://github.com/jenkins-infra/helpdesk/issues/4124#issuecomment-2275506395

Added jdk-17 and jdk-21 Ubuntu-22(AMD and ARM) ssh agents to infra.ci, Ubuntu-22 (AMD and ARM) template will have jdk-11 by default.

### CODE REVIEW

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 1]
```diff
@@ -288,11 +288,15 @@ controller:
                     EOF
                     systemctl daemon-reload
                     systemctl restart docker
+                    export DEFAULT_JDK=/opt/jdk-11
+                    update-alternatives --install /usr/bin/java java "${DEFAULT_JDK}/bin/java" 2000
+                    echo "JAVA_HOME=${DEFAULT_JDK}" >> /etc/environment
                     rm -f /etc/sudoers.d/90-cloud-init-users
                     echo "END CLOUDINIT"
                   javaPath: "/opt/jdk-17/bin/java"
+                  javaHome: "/opt/jdk-11"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:297]:
```suggestion
```

Not a JCasc valid attribute :)

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 2]
```diff
@@ -288,11 +288,15 @@ controller:
                     EOF
                     systemctl daemon-reload
                     systemctl restart docker
+                    export DEFAULT_JDK=/opt/jdk-11
+                    update-alternatives --install /usr/bin/java java "${DEFAULT_JDK}/bin/java" 2000
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:292]:
```suggestion
                    update-alternatives --install /usr/bin/java java "^${DEFAULT_JDK}/bin/java" 2000
```

Need to escape `$` sign on JCasc if you want it to be litteral

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 3]
```diff
@@ -288,11 +288,15 @@ controller:
...
+                    echo "JAVA_HOME=${DEFAULT_JDK}" >> /etc/environment
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:293]:
```suggestion
                    echo "JAVA_HOME=^${DEFAULT_JDK}" >> /etc/environment
```

Need to escape `$` sign on JCasc if you want it to be litteral

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 4]
```diff
@@ -305,13 +309,143 @@ controller:
                   storageAccountNameReferenceType: "existing"
                   storageAccountType: "Standard_LRS"
                   subnetName: "infra-ci-jenkins-io-sponsorship-vnet-ephemeral-agents"
-                  templateDesc: "Dynamically provisioned Ubuntu 22.04 LTS machine"
+                  templateDesc: "Dynamically provisioned Ubuntu 22.04 LTS AMD machine"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:312]:
```suggestion
                  templateDesc: "Dynamically provisioned Ubuntu 22.04 LTS x86_64 machine"
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 5]
```diff
@@ -305,13 +309,143 @@ controller:
...
+                    cat <<EOF >/etc/docker/daemon.json
+                    {
+                      "registry-mirrors": ["https://dockerhubmirror.azurecr.io"]
+                    }
+                    EOF
+                    systemctl daemon-reload
+                    systemctl restart docker
+                    export DEFAULT_JDK=/opt/jdk-17
+                    update-alternatives --install /usr/bin/java java "${DEFAULT_JDK}/bin/java" 2000
+                    echo "JAVA_HOME=${DEFAULT_JDK}" >> /etc/environment
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:358]:
```suggestion
                    update-alternatives --install /usr/bin/java java "^${DEFAULT_JDK}/bin/java" 2000
                    echo "JAVA_HOME=^${DEFAULT_JDK}" >> /etc/environment
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 6]
```diff
@@ -305,13 +309,143 @@ controller:
...
+                    rm -f /etc/sudoers.d/90-cloud-init-users
+                    echo "END CLOUDINIT"
+                  javaPath: "/opt/jdk-17/bin/java"
+                  javaHome: "/opt/jdk-17"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:362]:
```suggestion
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 7]
```diff
@@ -305,13 +309,143 @@ controller:
...
+                  noOfParallelJobs: 1
+                  osDiskSize: 150
+                  osDiskStorageAccountType: "Standard_LRS"
+                  osType: "Linux"
+                  retentionStrategy: "azureVMCloudOnce"
+                  spotInstance: false
+                  storageAccountNameReferenceType: "existing"
+                  storageAccountType: "Standard_LRS"
+                  subnetName: "infra-ci-jenkins-io-sponsorship-vnet-ephemeral-agents"
+                  templateDesc: "Dynamically provisioned Ubuntu 22.04 maven-17 LTS AMD machine"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:377]:
```suggestion
                  templateDesc: "Dynamically provisioned Ubuntu 22.04 maven-17 LTS x86_64 machine"
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 8]
```diff
@@ -305,13 +309,143 @@ controller:
...
+                    cat <<EOF >/etc/docker/daemon.json
+                    {
+                      "registry-mirrors": ["https://dockerhubmirror.azurecr.io"]
+                    }
+                    EOF
+                    systemctl daemon-reload
+                    systemctl restart docker
+                    export DEFAULT_JDK=/opt/jdk-21
+                    update-alternatives --install /usr/bin/java java "${DEFAULT_JDK}/bin/java" 2000
+                    echo "JAVA_HOME=${DEFAULT_JDK}" >> /etc/environment
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:423]:
```suggestion
                    update-alternatives --install /usr/bin/java java "^${DEFAULT_JDK}/bin/java" 2000
                    echo "JAVA_HOME=^${DEFAULT_JDK}" >> /etc/environment
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 9]
```diff
@@ -305,13 +309,143 @@ controller:
...
+                    rm -f /etc/sudoers.d/90-cloud-init-users
+                    echo "END CLOUDINIT"
+                  javaPath: "/opt/jdk-17/bin/java"
+                  javaHome: "/opt/jdk-21"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:427]:
```suggestion
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 10]
```diff
@@ -305,13 +309,143 @@ controller:
...
+                  noOfParallelJobs: 1
+                  osDiskSize: 150
+                  osDiskStorageAccountType: "Standard_LRS"
+                  osType: "Linux"
+                  retentionStrategy: "azureVMCloudOnce"
+                  spotInstance: false
+                  storageAccountNameReferenceType: "existing"
+                  storageAccountType: "Standard_LRS"
+                  subnetName: "infra-ci-jenkins-io-sponsorship-vnet-ephemeral-agents"
+                  templateDesc: "Dynamically provisioned Ubuntu 22.04 maven-21 LTS AMD machine"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:442]:
```suggestion
                  templateDesc: "Dynamically provisioned Ubuntu 22.04 maven-21 LTS x86_64 machine"
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 11]
```diff
@@ -443,11 +577,15 @@ controller:
                     EOF
                     systemctl daemon-reload
                     systemctl restart docker
+                    export DEFAULT_JDK=/opt/jdk-11
+                    update-alternatives --install /usr/bin/java java "${DEFAULT_JDK}/bin/java" 2000
+                    echo "JAVA_HOME=${DEFAULT_JDK}" >> /etc/environment
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:582]:
```suggestion
                    update-alternatives --install /usr/bin/java java "^${DEFAULT_JDK}/bin/java" 2000
                    echo "JAVA_HOME=^${DEFAULT_JDK}" >> /etc/environment
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 12]
```diff
@@ -443,11 +577,15 @@ controller:
...
                     rm -f /etc/sudoers.d/90-cloud-init-users
                     echo "END CLOUDINIT"
                   javaPath: "/opt/jdk-17/bin/java"
+                  javaHome: "/opt/jdk-11"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:586]:
```suggestion
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 13]
```diff
@@ -443,11 +577,15 @@ controller:
...
                   jvmOptions: "-XX:+PrintCommandLineFlags"
-                  labels: "linux-arm64 linux-arm64-docker azure-vms-jenkins-sponsorship arm64linux"
+                  labels: "linux-arm64 linux-arm64-docker azure-vms-jenkins-sponsorship arm64linux jdk11 maven-11"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:588]: Is it the default to use? Or the amd64? ;)

**Review by dduportal — CHANGES_REQUESTED** [review]
First batch of reviews

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 14]
```diff
@@ -305,13 +309,143 @@ controller:
...
+                    mkdir -p /etc/docker
+                    cat <<EOF >/etc/docker/daemon.json
+                    {
+                      "registry-mirrors": ["https://dockerhubmirror.azurecr.io"]
+                    }
+                    EOF
+                    systemctl daemon-reload
+                    systemctl restart docker
+                    export DEFAULT_JDK=/opt/jdk-17
+                    update-alternatives --install /usr/bin/java java "${DEFAULT_JDK}/bin/java" 2000
```
Comment (jayfranco999) [inline config/jenkins_infra.ci.jenkins.io.yaml:357]:
```suggestion
                    update-alternatives --install /usr/bin/java java "^${DEFAULT_JDK}/bin/java" 2000
```

File: config/jenkins_infra.ci.jenkins.io.yaml (hunk 13 above)
Comment (jayfranco999) [inline config/jenkins_infra.ci.jenkins.io.yaml:588]:
```suggestion
                  labels: "linux-arm64 linux-arm64-docker azure-vms-jenkins-sponsorship arm64linux jdk11-arm64 maven-11-arm-64"
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 15]
```diff
@@ -467,6 +605,136 @@ controller:
...
```
Comment (jayfranco999) [inline config/jenkins_infra.ci.jenkins.io.yaml:646]:
```suggestion
                    update-alternatives --install /usr/bin/java java "^${DEFAULT_JDK}/bin/java" 2000
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 16]
```diff
@@ -467,6 +605,136 @@ controller:
...
+                    echo "JAVA_HOME=${DEFAULT_JDK}" >> /etc/environment
+                    rm -f /etc/sudoers.d/90-cloud-init-users
+                    echo "END CLOUDINIT"
+                  javaPath: "/opt/jdk-17/bin/java"
+                  javaHome: "/opt/jdk-17"
```
Comment (jayfranco999) [inline config/jenkins_infra.ci.jenkins.io.yaml:651]:
```suggestion
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 17]
```diff
@@ -264,15 +264,15 @@ controller:
                     galleryImageVersion: "1.86.0"
                     galleryName: "prod_packer_images"
                     galleryResourceGroup: "prod-packer-images"
-                    gallerySubscriptionId: "${JENKINSINFRA_AZURE_SECONDARY_SUBSCRIPTION_ID}"
+                    gallerySubscriptionId: "^${JENKINSINFRA_AZURE_SECONDARY_SUBSCRIPTION_ID}"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:267]: What is the reason of this change @jayfranco999 ? it is unrelated to your PR and will break things 🤔

Comment (jayfranco999) [inline config/jenkins_infra.ci.jenkins.io.yaml:267]:
```suggestion
                    gallerySubscriptionId: "${JENKINSINFRA_AZURE_SECONDARY_SUBSCRIPTION_ID}"
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 18]
```diff
@@ -264,15 +264,15 @@ controller:
...
                   imageTopLevelType: "advanced"
                   initScript: |-
                     #!/bin/bash
                     set -eux
                     echo "START CLOUDINIT"
                     systemctl stop datadog-agent.service
                     mkdir -p /var/log/datadog /etc/datadog-agent
-                    sed 's/api_key:.*/api_key: ${JENKINS_CI_DATADOG_API_KEY}/' /etc/datadog-agent/datadog.yaml.example > /etc/datadog-agent/datadog.yaml
+                    sed 's/api_key:.*/api_key: ^${JENKINS_CI_DATADOG_API_KEY}/' /etc/datadog-agent/datadog.yaml.example > /etc/datadog-agent/datadog.yaml
```
Comment (jayfranco999) [inline config/jenkins_infra.ci.jenkins.io.yaml:275]:
```suggestion
                    sed 's/api_key:.*/api_key: ${JENKINS_CI_DATADOG_API_KEY}/' /etc/datadog-agent/datadog.yaml.example > /etc/datadog-agent/datadog.yaml
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 19]
```diff
@@ -288,14 +288,17 @@ controller:
...
                     rm -f /etc/sudoers.d/90-cloud-init-users
                     echo "END CLOUDINIT"
                   javaPath: "/opt/jdk-17/bin/java"
                   jvmOptions: "-XX:+PrintCommandLineFlags"
-                  labels: "linux-amd64 linux-amd64-docker azure-vms-jenkins-sponsorship"
+                  labels: "linux-amd64 linux-amd64-docker azure-vms-jenkins-sponsorship jdk11 maven-11"
                   location: "East US 2"
                   maxVirtualMachinesLimit: 50
-                  existingStorageAccountName: "${AZURE_VM_AGENTS_STORAGE_ACCOUNT_SECONDARY_SUBSCRIPTION}"
+                  existingStorageAccountName: "^${AZURE_VM_AGENTS_STORAGE_ACCOUNT_SECONDARY_SUBSCRIPTION}"
```
Comment (jayfranco999) [inline config/jenkins_infra.ci.jenkins.io.yaml:301]:
```suggestion
                  existingStorageAccountName: "${AZURE_VM_AGENTS_STORAGE_ACCOUNT_SECONDARY_SUBSCRIPTION}"
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 20]
```diff
@@ -292,11 +292,14 @@ controller:
...
                     systemctl restart docker
+                    export DEFAULT_JDK=/opt/jdk-11
+                    update-alternatives --install /usr/bin/java java "^${DEFAULT_JDK}/bin/java" 2000
+                    echo "JAVA_HOME=^${DEFAULT_JDK}" >> /etc/environment
                     rm -f /etc/sudoers.d/90-cloud-init-users
                     echo "END CLOUDINIT"
                   javaPath: "/opt/jdk-17/bin/java"
                   jvmOptions: "-XX:+PrintCommandLineFlags"
-                  labels: "linux-amd64 linux-amd64-docker azure-vms-jenkins-sponsorship"
+                  labels: "linux-amd64 linux-amd64-docker azure-vms-jenkins-sponsorship jdk11 maven-11"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:302]:
```suggestion
                  labels: "linux-amd64 linux-amd64-docker azure-vms-jenkins-sponsorship"
```

We discussed about using the ARM64 as default in this controller (we like to live on the edge ;) )

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 21]
```diff
@@ -447,11 +578,14 @@ controller:
...
                     systemctl restart docker
+                    export DEFAULT_JDK=/opt/jdk-11
+                    update-alternatives --install /usr/bin/java java "^${DEFAULT_JDK}/bin/java" 2000
+                    echo "JAVA_HOME=^${DEFAULT_JDK}" >> /etc/environment
                     rm -f /etc/sudoers.d/90-cloud-init-users
                     echo "END CLOUDINIT"
                   javaPath: "/opt/jdk-17/bin/java"
                   jvmOptions: "-XX:+PrintCommandLineFlags"
-                  labels: "linux-arm64 linux-arm64-docker azure-vms-jenkins-sponsorship arm64linux"
+                  labels: "linux-arm64 linux-arm64-docker azure-vms-jenkins-sponsorship arm64linux jdk11-arm64 maven-11-arm64"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:588]:
```suggestion
                  labels: "linux-arm64 linux-arm64-docker azure-vms-jenkins-sponsorship arm64linux jdk11 maven-11 ubuntu-22-arm64 ubuntu-arm64"
```

- Add the default JDK11 labels for existing pipelines (ref. https://github.com/jenkins-infra/kubernetes-management/pull/5526/files#r1716774585)
- Add the templateName as explicit label
- Add an ubuntu "non-versioned" default label

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 22]
```diff
@@ -309,13 +312,141 @@ controller:
...
+                  storageAccountNameReferenceType: "existing"
+                  storageAccountType: "Standard_LRS"
+                  subnetName: "infra-ci-jenkins-io-sponsorship-vnet-ephemeral-agents"
+                  templateDesc: "Dynamically provisioned Ubuntu 22.04 maven-21 LTS x86_64 machine"
+                  templateName: "ubuntu-22-amd64-maven-21"
+                  usageMode: NORMAL
+                  usePrivateIP: true
+                  virtualMachineSize: "Standard_D4ads_v5" # 4 vCPUS / 16 Gb / Max 150 Gb local storage
+                  virtualNetworkName: "infra-ci-jenkins-io-sponsorship-vnet"
+                  virtualNetworkResourceGroupName: "infra-ci-jenkins-io-sponsorship"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:449]:
```suggestion
```

Let's try to use only arm64, and we keep 1 x86_64 for backward compatibility

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 23]
```diff
@@ -464,13 +598,141 @@ controller:
...
+                    systemctl daemon-reload
+                    systemctl restart docker
+                    export DEFAULT_JDK=/opt/jdk-17
+                    update-alternatives --install /usr/bin/java java "^${DEFAULT_JDK}/bin/java" 2000
+                    echo "JAVA_HOME=^${DEFAULT_JDK}" >> /etc/environment
+                    rm -f /etc/sudoers.d/90-cloud-init-users
+                    echo "END CLOUDINIT"
+                  javaPath: "/opt/jdk-17/bin/java"
+                  jvmOptions: "-XX:+PrintCommandLineFlags"
+                  labels: "linux-arm64-maven-17"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:652]:
```suggestion
                  labels: "linux-arm64-maven-17 jdk17 maven-17 ubuntu-22-arm64-maven-17 ubuntu-arm64-maven-17"
```

File: config/jenkins_infra.ci.jenkins.io.yaml [hunk 24]
```diff
@@ -464,13 +598,141 @@ controller:
...
+                    systemctl daemon-reload
+                    systemctl restart docker
+                    export DEFAULT_JDK=/opt/jdk-21
+                    update-alternatives --install /usr/bin/java java "^${DEFAULT_JDK}/bin/java" 2000
+                    echo "JAVA_HOME=^${DEFAULT_JDK}" >> /etc/environment
+                    rm -f /etc/sudoers.d/90-cloud-init-users
+                    echo "END CLOUDINIT"
+                  javaPath: "/opt/jdk-17/bin/java"
+                  jvmOptions: "-XX:+PrintCommandLineFlags"
+                  labels: "linux-arm64-maven-21"
```
Comment (dduportal) [inline config/jenkins_infra.ci.jenkins.io.yaml:716]:
```suggestion
                  labels: "linux-arm64-maven-21 jdk21 maven-21 ubuntu-22-arm64-maven-21 ubuntu-arm64-maven-21"
```

**Review by dduportal — APPROVED** [review]
LGTM! Helm file diff looks good!

**Pull request merged**